src/v/cluster/feature_manager.cc

```diff
@@ -287,7 +293,61 @@ ss::future<> feature_manager::maybe_log_license_check_info() {
     }
 }
 
+bool feature_manager::need_to_verify_enterprise_license() {
+    return features::is_major_version_upgrade(
+      _feature_table.local().get_active_version(),
+      _feature_table.local().get_latest_logical_version());
+}
+
+void feature_manager::verify_enterprise_license() {
+    vlog(clusterlog.info, "Verifying enterprise license...");
+
+    if (!need_to_verify_enterprise_license()) {
+        vlog(clusterlog.info, "Enterprise license verification skipped...");
+        _verified_enterprise_license.signal();
+        return;
+    }
+
+    const auto& license = _feature_table.local().get_license();
+    auto license_missing_or_expired = !license || license->is_expired();
+    auto enterprise_features = report_enterprise_features();
+
+    vlog(
+      clusterlog.info,
+      "Verifying enterprise license: active_version={}, latest_version={}, "
+      "enterprise_features=[{}], license_missing_or_expired={}",
+      _feature_table.local().get_active_version(),
+      _feature_table.local().get_latest_logical_version(),
+      enterprise_features.enabled(),
+      license_missing_or_expired);
+
+    if (enterprise_features.any() && license_missing_or_expired) {
+        throw std::runtime_error{fmt::format(
+          "A Redpanda Enterprise Edition license is required to use enterprise "
+          "features: ([{}]). Enter an active license key (for example, rpk "
+          "cluster license set <key>). To request a license, see "
+          "https://redpanda.com/license-request. For more information, see "
+          "https://docs.redpanda.com/current/get-started/licenses.",
+          enterprise_features.enabled())};
+    }
+
+    _verified_enterprise_license.signal();
+}
+
 ss::future<> feature_manager::maybe_update_feature_table() {
+    // Before doing any feature enablement or active version update, check that
+    // the cluster has an enterprise license if they use enterprise features
+    if (need_to_verify_enterprise_license()) {
```

Comment:
do you need to check `need_to_verify_enterprise_license` here, if you have code elsewhere that already checks it, and sets `_verified_enterprise_license`?

Reply:
It's not required, but it means that we can unblock this fibre (slightly) earlier for non-upgrading nodes. I went with this to make startup more similar to the existing behaviour for non-upgrading nodes, but it is not strictly necessary.